Improve rendering flow

johanjanssens opened the pull request:
This PR changes the rendering flow in the framework and allows controller action methods to return content without needing to render a view.

- If the controller returns a string or a object that can be casted to a string return the result of the controller action will be send to the client. If not the will be asked to try and render the result.

- If the controller returns a KObjectConfigFormat object. The content type will automatically set based on the media type of object. If not the implementing code is responsible for setting it.

- In case of a view content type negotiation happens in the request. A view does no longer need to specifically define it's content type. 

- A new isForwarded() method has been added to the dispatcher to check if a dispatcher was forwarded or not, a forwarded dispatcher will not send a response, it will return it instead. 

This changes makes the view into a none required part of the MVC triad. The view will only be called upon to render complex data structures that cannot be casted to a string. Simple array data structures do not require a view for rendering.

This also extends on the concepts introduced in https://github.com/joomlatools/joomlatools-framework/pull/184 and allows a controller action method to return a string or a object that can be casted to a string. 

### Backwards compatibility

Due the changes in implementation if a controller is rendered, without being dispatched, it will no longer set the rendered result in the response, instead the rendered result will be returned. It's up to the implementing code to set the result in the response.

ercanozkaya commented on code/libraries/joomlatools/library/controller/model.php:
```diff
@@ -179,8 +182,11 @@ protected function _actionRender(KControllerContextInterface $context)
         //Execute the action
-        if($this->execute($action, $context) !== false) {
-            $result = parent::_actionRender($context);
+        if($result = $this->execute($action, $context) !== false)
```
@johanjanssens If I understood the idea correctly the following should return directly `{"foo": "bar"}` right?
```
protected function _actionFoo(KControllerContextInterface $context)
    {
        return new KObjectConfigJson(['foo' => 'bar']);
    }
```

But this still calls read (or browse) and returns the result from that action instead of  the result of foo action

johanjanssens replied:
It might still not fully understand your use-case here, the code will not render a view if a string or object casteable to  string is returned from get or browse.

johanjanssens replied:
How are you calling actionFoo?

ercanozkaya replied:
By using an action override in the payload. Full POST data looks something like `{'csrf_token': 'token', '_action': 'foo', 'extra': 'data'}`

johanjanssens commented:
@ercanozkaya Added logic to always return a representation of the resource for POST and PUT requests. Similar to browse and read, if the add, or edit action returns a string or an object that can be casted to a string it will be returned, if not the view will be asked to render a representation of the resource.

ercanozkaya commented on code/libraries/joomlatools/library/dispatcher/behavior/resettable.php:
```diff
@@ -30,28 +31,28 @@ public function isSupported()
         $request = $mixer->getRequest();
 
-        if(!$request->isSafe() && !$request->isAjax() && $request->getFormat() == 'html') {
+        if(!$request->isSafe() && $request->getReferrer() && $request->getContentType() == 'application/x-www-form-urlencoded') {
```
@johanjanssens This breaks the AJAX responses. When you send an AJAX request on a browser you have a referrer and the content type by default is application/x-www-form-urlencoded. So the condition is fullfilled. Suggest to revert this one.

johanjanssens replied:
Fixed this. Added check back and also making sure referrer exists. Should work now.

ercanozkaya commented on code/libraries/joomlatools/library/dispatcher/http.php:
```diff
@@ -252,7 +252,16 @@ protected function _actionPost(KDispatcherContextInterface $context)
+            //Return the new representation of the resource
+            if ($context->response->isSuccess())
+            {
+                if(!is_string($result)  && !is_callable(array($result, '__toString'))) {
```
@johanjanssens is_callable return value is erratic on my end. For rowsets it returns true even though we don't have __toString. 
The following worked:
`if(!is_string($result)  && !(is_object($result) && method_exists($result, '__toString'))) {`
This is actually a safer call too. If $result is somehow an array in the original a notice will be generated

johanjanssens replied:
Agreed, I have changed this everywhere to keep the consistency.

ercanozkaya commented on code/libraries/joomlatools/library/dispatcher/http.php:
```diff
@@ -302,7 +311,16 @@ protected function _actionPut(KDispatcherContextInterface $context)
+            //Return the new representation of the resource
+            if ($context->response->isSuccess())
+            {
+                if(!is_string($result)  && !is_callable(array($result, '__toString'))) {
```
@johanjanssens same as above

johanjanssens replied:
Done.

johanjanssens commented:
@ercanozkaya Can you give this a final review. Addressed your feedback, and also removed the Csv view. Replaced it by ObjectConfigCsv. Don't need a Cvs view anywhere, can just return a csv object if ever needed.

ercanozkaya commented:
As mentioned on Slack, decoratable render() call is not having an effect in this branch. All other issues are now fixed

ercanozkaya commented:
@johanjanssens All good here
